multicluster/controllers/multicluster/common/controller_utils.go

```diff
@@ -78,6 +87,71 @@ func parseServiceCIDRFromError(msg string) (string, error) {
 	return match[1], nil
 }
 
+func isK8sVersionGreaterThanOrEqualTo(discoveryClient discovery.DiscoveryInterface, expectedVersion string) (bool, error) {
+	versionInfo, err := discoveryClient.ServerVersion()
+	if err != nil {
+		return false, err
+	}
+	vers, err := version.ParseGeneric(versionInfo.GitVersion)
+	if err != nil {
+		return false, err
+	}
+	return vers.AtLeast(version.MustParseGeneric(expectedVersion)), nil
+}
+
+func getClusterServiceCIDR(ctx context.Context, client client.Client) (string, error) {
+	// A ServiceCIDR CR named 'kubernetes' will be created by default since v1.33.0.
+	// We will retrieve it to get the Service CIDR.
+	var svcCIDR networkingv1.ServiceCIDR
+	err := client.Get(ctx, types.NamespacedName{Name: "kubernetes"}, &svcCIDR)
+	if err != nil {
+		return "", fmt.Errorf("failed to get ServiceCIDR 'kubernetes': %w", err)
+	}
+
+	for _, cidr := range svcCIDR.Spec.CIDRs {
+		if isIPv4CIDR(cidr) {
+			return cidr, nil
+		}
+	}
+	return "", fmt.Errorf("IPv4 Service CIDR not found")
+}
+
+func createDiscoveryClient(config *rest.Config) (discovery.DiscoveryInterface, error) {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return discoveryClient, nil
+}
+
+func DiscoverClusterServiceCIDR(ctx context.Context, mgrConfig *rest.Config, client client.Client, namespace string) (string, error) {
+	var versionMatched bool
+	var cidr string
+	discoveryClient, err := createDiscoveryClientFn(mgrConfig)
+	if err != nil {
+		return "", err
+	}
+	if versionMatched, err = isK8sVersionGreaterThanOrEqualTo(discoveryClient, "v1.33.0"); versionMatched {
+		cidr, err = getClusterServiceCIDR(ctx, client)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		klog.InfoS("Failed to get the K8s version, falling back to legacy Service CIDR detection",
```

Comment:
Here not necessarily isK8sVersionGreaterThanOrEqualTo returns an error, but it may just be K8s version is old.

Reply:
Refined a few codes, @jianjuns could you take a look again? thanks.